sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -88,5 +81,146 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
                 ProcessNext(message, pipeline);
             }
         }
+
+        private async ValueTask<string?> GetHeaderValueFromCredentialAsync(HttpMessage message, bool async)
+        {
+            try
+            {
+                var requestContext = new TokenRequestContext(_scopes, message.Request.ClientRequestId);
+                AccessToken token = async
+                    ? await _credential.GetTokenAsync(requestContext, message.CancellationToken).ConfigureAwait(false)
+                    : _credential.GetToken(requestContext, message.CancellationToken);
+
+                return _accessTokenCache.SaveToken(token);
+            }
+            catch (Exception e)
+            {
+                _accessTokenCache.ResetToken(e);
```

Comment:
How often would we try to refresh the background token? Do we want to try and refresh in the background without blocking the current request at all?

Reply:
This code is unreachable when token hasn't expired. I've added test for that case.

Reply:
@pakrym this is what we decided to do in JS. Essentially we queue the refresh when the first request comes in after we've reached the refresh buffer, but we don't block that request. We could use the same functionality.

Reply:
Added another test, fixed the issue

Reply:
The current implementation still does token refresh inline with a request or am I missing something?

Reply:
I think it is. Previously, if request has failed, existing token will be erased even if it isn't expired. With this change, it will be reset only if it has already expired, otherwise cache will be returned in the `TokenState.Valid` so that next request can try to refresh token again:

```cs
public void ResetTokenIfExpired(Exception exception)
{
string? headerValue = default;
    TaskCompletionSource<string?>? pendingTcs;

    lock (_syncObj)
    {
        if (DateTimeOffset.UtcNow >= _expiresOn)
        {
            _headerValue = default;
            _refreshOn = default;
            _expiresOn = default;
            _tokenState = TokenState.Invalid;

            pendingTcs = _pendingTcs;
            _pendingTcs = null;
        }
        else if (_tokenState == TokenState.AboutToExpire)
        {
            _tokenState = TokenState.Valid;
            if (_pendingTcs != default)
            {
                exception = new InvalidOperationException($"{nameof(GetHeaderValueAsync)} shouldn't wait in {TokenState.AboutToExpire} state.", exception);
            }
        }
    }

    if (headerValue != null)
    {
        pendingTcs?.SetResult(headerValue);
    }
    else
    {
        pendingTcs?.SetException(exception);
    }
}
```

Here is the situation we handle: Two options are possible: If at the time of request failure

Reply:
Yes, but trying to get a new token would still happen inline with sending the current request not on a background thread.

Reply:
Fixed